packages/adminrouter/extra/src/test-harness/tests/test_service.py

```diff
@@ -553,3 +554,198 @@ def test_if_scheme_is_honoured_in_mesos_scheduler_entry(
         assert resp.status_code == 200
         req_data = resp.json()
         assert req_data['endpoint_id'] == "https://127.0.0.4:443"
+
+    def test_if_ar_with_empty_cache_waits_for_marathon_during_service_resolve(
+            self, mocker, nginx_class, valid_user_header):
+        # Make service endpoint resolve only Marathon-related data:
+        mocker.send_command(
+            endpoint_id='http://127.0.0.2:5050',
+            func_name='set_frameworks_response',
+            aux_data=[])
+        mocker.send_command(
+            endpoint_id='http://127.0.0.1:8123',
+            func_name='set_srv_response',
+            aux_data=EMPTY_SRV)
+
+        # Make Mock endpoint stall a little, make sure AR cache update timeouts
+        # are big enough to swallow it:
+        backend_request_timeout = 6
+        refresh_lock_timeout = backend_request_timeout * 2
+
+        # Make period cache refreshes so rare that they do not get into
+        # picture:
+        ar = nginx_class(cache_first_poll_delay=1200,
+                         cache_poll_period=1200,
+                         cache_expiration=1200,
+                         cache_max_age_soft_limit=1200,
+                         cache_max_age_hard_limit=1800,
+                         cache_backend_request_timeout=backend_request_timeout,
+                         cache_refresh_lock_timeout=refresh_lock_timeout,
+                         )
+        url = ar.make_url_from_path("/service/scheduler-alwaysthere/")
+
+        mocker.send_command(endpoint_id='http://127.0.0.1:8080',
+                            func_name='always_stall',
+                            aux_data=backend_request_timeout * 0.5)
+
+        # Measure the time it took and the results:
+        with GuardedSubprocess(ar):
+            t_start = time.time()
+            resp = requests.get(url,
+                                allow_redirects=False,
+                                headers=valid_user_header)
+
+            t_spent = time.time() - t_start
+            assert resp.status_code == 200
+            data = resp.json()
+            assert data['endpoint_id'] == 'http://127.0.0.1:16000'
+
+            # If AR wait for cache during resolve, then time spent should be
+            # greater than the stall time that has been set. Due to the fact
+            # that update coroutines are not separated yet, this will be
```

Comment:
> I don't know what update coroutines are.
NGINX lightweight threads, I had documentation somewhere but I cannot remember where now :(

> Would it be correct to remove the comment about this and change the assertion to >=?
No, because we configure Marathon mock to delay for `backend_request_timeout * 0.5`, if the time spent is exaclty `backend_request_timeout * 0.5` then it means that the request returned earlier than `backend_request_timeout * 0.5` as nothing in instantaneous and there is still some context switching and data processing. It is much more reliable this way.